actors/builtin/market/market_test.go

```diff
@@ -1 +1,252 @@
+			WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID)
+
+		rt := builder.Build(t)
+
+		actor := marketActorTestHarness{t: t}
+		actor.constructAndVerify(rt)
+
+		return rt, &actor
+	}
+
+	t.Run("adds to provider escrow funds", func(t *testing.T) {
+		rt, actor := setup()
+
+		testCases := []struct {
+			delta int64
+			total int64
+		}{
+			{10, 10},
+			{20, 30},
+			{40, 70},
+		}
+
+		for _, tc := range testCases {
+			rt.SetCaller(provider, builtin.StorageMinerActorCodeID)
+			rt.SetReceived(abi.NewTokenAmount(tc.delta))
+			actor.expectProviderControlAddress(rt, provider, worker)
+
+			rt.Call(actor.AddBalance, &provider)
+
+			rt.Verify()
+
+			rt.GetState(&st)
+			assert.Equal(t, abi.NewTokenAmount(tc.total), st.GetEscrowBalance(rt, provider))
+		}
+	})
+
+	t.Run("adds to non-provider escrow funds", func(t *testing.T) {
+		rt, actor := setup()
+
+		testCases := []struct {
+			delta int64
+			total int64
+		}{
+			{10, 10},
+			{20, 30},
+			{40, 70},
+		}
+
+		for _, tc := range testCases {
+			rt.SetCaller(client, builtin.AccountActorCodeID)
+			rt.SetReceived(abi.NewTokenAmount(tc.delta))
+			rt.ExpectValidateCallerType(builtin.CallerTypesSignable...)
+
+			rt.Call(actor.AddBalance, &client)
+
+			rt.Verify()
+
+			rt.GetState(&st)
+			assert.Equal(t, abi.NewTokenAmount(tc.total), st.GetEscrowBalance(rt, client))
+		}
+	})
+
+}
+
+func TestWithdrawBalance(t *testing.T) {
+	actor := marketActorTestHarness{t: t}
+	marketActor := tutil.NewIDAddr(t, 100)
+	provider := tutil.NewIDAddr(t, 101)
+	worker := tutil.NewIDAddr(t, 102)
+	client := tutil.NewIDAddr(t, 103)
+
+	var st market.State
+
+	builder := mock.NewBuilder(context.Background(), marketActor).
+		WithBalance(abi.NewTokenAmount(100), abi.NewTokenAmount(0)).
+		WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID)
+
+	t.Run("fails with a negative withdraw amount", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		params := market.WithdrawBalanceParams{
+			ProviderOrClientAddress: provider,
+			Amount:                  abi.NewTokenAmount(-1),
+		}
+
+		rt.ExpectAbort(exitcode.ErrIllegalArgument, func() {
+			rt.Call(actor.WithdrawBalance, &params)
+		})
+
+		rt.Verify()
+	})
+
+	t.Run("withdraws from provider escrow funds", func(t *testing.T) {
+		rt := builder.WithBalance(abi.NewTokenAmount(100), abi.NewTokenAmount(0)).Build(t)
+		actor.constructAndVerify(rt)
+		actor.addProviderFunds(rt, provider, worker, abi.NewTokenAmount(20))
+
+		rt.GetState(&st)
+		assert.Equal(t, abi.NewTokenAmount(20), st.GetEscrowBalance(rt, provider))
+
+		rt.SetCaller(provider, builtin.StorageMinerActorCodeID)
+		actor.expectProviderControlAddress(rt, provider, worker)
+		rt.ExpectSend(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, abi.NewTokenAmount(0), nil, exitcode.Ok)
+
+		withdrawAmount := abi.NewTokenAmount(1)
+		rt.ExpectSend(provider, builtin.MethodSend, nil, withdrawAmount, nil, exitcode.Ok)
```

Comment:
This must send to the owner address (please use distinct owner/worker addresses so this can be asserted).